openhcl/page_pool_alloc/src/lib.rs

```diff
@@ -23,6 +23,261 @@ use std::sync::Arc;
+            match state {
+                State::Free {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                } => InnerState::Free {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                },
+                State::Allocated {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                } => InnerState::Allocated {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                },
+                State::AllocatedPendingRestore { .. } => {
+                    panic!("should not save AllocatedPendingRestore")
+                }
+                State::Leaked {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                } => InnerState::Leaked {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                },
+            }
+        }
+    }
+
+    impl From<InnerState> for State {
+        fn from(state: InnerState) -> Self {
+            match state {
+                InnerState::Free {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                } => State::Free {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                },
+                InnerState::Allocated {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                } => State::AllocatedPendingRestore {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                },
+                InnerState::Leaked {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                } => State::Leaked {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                },
+            }
+        }
+    }
+
+    // TODO: It seems unfortunate to define this type which is the same as
+    // MemoryRangeWithNode in vm_topology.
+    #[derive(Protobuf)]
+    #[mesh(package = "openhcl.pagepool")]
+    struct MemoryRangeWithNode {
+        #[mesh(1)]
+        range: MemoryRange,
+        #[mesh(2)]
+        vnode: u32,
+    }
+
+    #[derive(Protobuf)]
+    #[mesh(package = "openhcl.pagepool")]
+    enum DeviceIdState {
```

Comment:
The device ID is just a compression technique. We shouldn't encode it into the saved state so literally. I think you just need to save a `Vec<DeviceId>`, `struct DeviceId { id: value, name: String }`. Don't even worry about what the integer id is, reallocate it using whatever scheme you like on restore.

Reply:
yeah i reworked this - we just save the strings everywhere, then do a vec `position` find. It's not the most efficient but realistically i'm not sure a hashmap will be faster given how few devices we'll have? 

If you really want a hashmap on restore, i can go ahead and do that.